File: coding-interview-linked-lists-plus/main_algorithms/remove_nth_from_end.py
# main_algorithms/remove_nth_from_end.py

from utils.linked_list_utils import ListNode

class SolutionRemoveNthFromEnd:
    """
    Problem: Remove Nth Node From End of List
    Given the head of a linked list, remove the nth node from the end of the list
    and return its head.

    Example 1:
    Input: head = [1,2,3,4,5], n = 2
    Output: [1,2,3,5]

    Example 2:
    Input: head = [1], n = 1
    Output: []

    Example 3:
    Input: head = [1,2], n = 1
    Output: [1]
    """

    def removeNthFromEnd_two_pass(self, head: ListNode, n: int) -> ListNode:
        """
        Approach 1: Two-pass Algorithm
        First pass: Calculate the length of the list.
        Second pass: Traverse to the (length - n)th node and remove the next node.

        Algorithm:
        1. Edge Case: If the list is empty, return None.
        2. First Pass (Calculate Length):
           a. Initialize a `dummy_head` node and point its `next` to `head`.
              This simplifies handling cases where the head itself needs to be removed.
           b. Traverse the list from `dummy_head` to find its total length.
           c. `length = 0` (or 1 if starting from head and incrementing from there).
              Keep traversing until `curr` is `None`.
        3. Calculate the position of the node to remove from the beginning:
           `pos_to_remove = length - n`.
           If `pos_to_remove` is 0, it means we need to remove the original head.
        4. Second Pass (Remove Node):
           a. Reset `current` pointer to `dummy_head`.
           b. Traverse `pos_to_remove` steps from `dummy_head`. The `current` pointer
              will then be at the node *before* the one to be removed.
           c. Perform the removal: `current.next = current.next.next`.
        5. Return `dummy_head.next`.

        Complexity Analysis:
        - Time Complexity: O(L), where L is the length of the linked list.
          We traverse the list once to find the length (O(L)) and once again
          to find the (L-n)th node (O(L)). Total is O(L).
        - Space Complexity: O(1). We only use a few extra pointers.
        """
        if not head:
            return None

        dummy_head = ListNode(0)
        dummy_head.next = head

        length = 0
        current = head
        while current:
            length += 1
            current = current.next
        
        # The node to remove is at index `length - n` (0-indexed)
        # We need to stop at the node before it, i.e., at index `length - n - 1`.
        # So we iterate `length - n` times from dummy_head.
        steps_to_predecessor = length - n
        
        if steps_to_predecessor < 0:
            # This case means n > length, which per problem constraints usually doesn't happen,
            # but good to be aware. For Nth from end, N is guaranteed to be valid.
            return head # Or raise error

        current = dummy_head
        for _ in range(steps_to_predecessor):
            current = current.next
        
        # current is now the node *before* the one to be removed
        current.next = current.next.next

        return dummy_head.next

    def removeNthFromEnd_one_pass(self, head: ListNode, n: int) -> ListNode:
        """
        Approach 2: One-pass Algorithm (Two Pointers)
        This method uses two pointers, `fast` and `slow`, separated by `n` nodes.
        When `fast` reaches the end of the list, `slow` will be at the node
        *before* the one to be removed.

        Algorithm:
        1. Edge Case: If the list is empty, return None.
        2. Create a `dummy_head` node and point its `next` to `head`.
           This simplifies handling cases where the head itself needs to be removed.
        3. Initialize `fast` and `slow` pointers, both starting at `dummy_head`.
        4. Move `fast` pointer `n+1` steps forward. This creates a gap of `n` nodes
           between `fast` and `slow`. (The `+1` is because `slow` needs to stop
           *before* the node to be removed).
        5. Now, move both `fast` and `slow` one step at a time until `fast` reaches `None`.
           When `fast` becomes `None`, `slow` will be at the node just before the `n`th
           node from the end.
        6. Perform the removal: `slow.next = slow.next.next`.
        7. Return `dummy_head.next`.

        Example: head = [1,2,3,4,5], n = 2
        dummy -> 1 -> 2 -> 3 -> 4 -> 5 -> None
        slow = dummy, fast = dummy

        Move fast n+1 = 3 steps:
        fast: dummy -> 1 -> 2 -> 3
        slow: dummy

        Move fast and slow together until fast is None:
        Iteration 1:
        slow: dummy -> 1
        fast: dummy -> 1 -> 2 -> 3 -> 4

        Iteration 2:
        slow: dummy -> 1 -> 2
        fast: dummy -> 1 -> 2 -> 3 -> 4 -> 5

        Iteration 3:
        slow: dummy -> 1 -> 2 -> 3
        fast: dummy -> 1 -> 2 -> 3 -> 4 -> 5 -> None
        Now fast is None. slow is at 3. The node to remove is 4.
        slow.next (4) = slow.next.next (5).
        Result: dummy -> 1 -> 2 -> 3 -> 5

        Complexity Analysis:
        - Time Complexity: O(L), where L is the length of the linked list.
          Both pointers traverse the list, with `fast` making roughly `L` steps
          and `slow` making `L-n` steps. Total is O(L).
        - Space Complexity: O(1). We only use a few extra pointers.
        """
        if not head:
            return None

        dummy_head = ListNode(0)
        dummy_head.next = head

        slow = dummy_head
        fast = dummy_head

        # Move fast pointer n+1 steps ahead
        # This ensures 'slow' is at the node *before* the one to be removed
        # when 'fast' reaches the end.
        for _ in range(n + 1):
            if not fast: # Should not happen if n is valid per constraints
                raise ValueError("n is larger than list length")
            fast = fast.next
        
        # Move both fast and slow until fast reaches the end
        while fast:
            slow = slow.next
            fast = fast.next
        
        # At this point, slow is the node BEFORE the one to be removed
        # e.g., if we want to remove 4, slow is at 3.
        # So, slow.next is the node to be removed.
        slow.next = slow.next.next

        return dummy_head.next

File: leetcode-linked-lists-solutions-complete/src/problems/removeNthFromEnd.ts
import { ListNode } from '@data-structures/ListNode';

/**
 * Problem: Remove Nth Node From End of List
 * Given the head of a linked list, remove the n-th node from the end of the list and return its head.
 */

/**
 * Solution 1: One-Pass (Two-Pointer Approach)
 *
 * This is the optimal approach, solving the problem in a single pass.
 * It uses two pointers, `fast` and `slow`, separated by `n` nodes.
 *
 * 1. Create a dummy node and point its `next` to `head`. This handles edge cases like removing
 *    the head of the list or an empty list.
 * 2. Initialize `fast` and `slow` pointers to the `dummy` node.
 * 3. Move `fast` `n` steps ahead of `slow`. Now, `fast` is `n` nodes ahead of `slow`.
 * 4. Move both `fast` and `slow` one step at a time until `fast` reaches the end of the list (i.e., `fast.next` is `null`).
 *    At this point, `slow` will be pointing to the node *before* the `n`-th node from the end.
 * 5. Remove the `n`-th node from the end: `slow.next = slow.next.next`.
 * 6. Return `dummy.next` as the head of the modified list.
 *
 * @param head The head of the linked list.
 * @param n The position from the end to remove (1-indexed).
 * @returns The head of the modified linked list.
 *
 * Time Complexity: O(L) where L is the number of nodes in the list.
 *                  We traverse the list a single time.
 * Space Complexity: O(1) as we only use a few extra pointers.
 */
export function removeNthFromEndOnePass(head: ListNode | null, n: number): ListNode | null {
    // Create a dummy node. This helps in cases where the head itself needs to be removed.
    const dummy = new ListNode(0);
    dummy.next = head;

    let fast: ListNode | null = dummy;
    let slow: ListNode | null = dummy;

    // Move fast pointer n steps ahead
    for (let i = 0; i <= n; i++) { // move n+1 steps, so fast is n nodes ahead of slow's target parent
        if (fast === null) {
            // This case should ideally not happen based on problem constraints (1 <= n <= sz)
            // but is a good defensive check for invalid 'n'
            return head;
        }
        fast = fast.next;
    }

    // Now, move both fast and slow pointers until fast reaches the end.
    // When fast reaches null, slow will be at the node *before* the one to be removed.
    while (fast !== null) {
        slow = slow!.next;
        fast = fast.next;
    }

    // `slow.next` is the node to be removed.
    // Bypass it by setting `slow.next` to `slow.next.next`.
    if (slow && slow.next) {
        slow.next = slow.next.next;
    }

    return dummy.next; // Return the head of the modified list (skipping the dummy node)
}

/**
 * Solution 2: Two-Pass Approach
 *
 * This approach solves the problem by first finding the length of the list,
 * and then traversing again to find the node to remove.
 *
 * 1. Create a dummy node and point its `next` to `head`.
 * 2. First pass: Iterate through the list to count the total number of nodes (`length`).
 * 3. Calculate the position of the node to remove from the beginning: `nodeToRemoveFromStart = length - n`.
 * 4. Second pass: Iterate from the `dummy` node up to `nodeToRemoveFromStart` steps.
 *    The pointer `current` will then be at the node *before* the one to be removed.
 * 5. Remove the node: `current.next = current.next.next`.
 * 6. Return `dummy.next`.
 *
 * @param head The head of the linked list.
 * @param n The position from the end to remove (1-indexed).
 * @returns The head of the modified linked list.
 *
 * Time Complexity: O(L) where L is the number of nodes in the list.
 *                  We traverse the list twice: once to count, once to remove.
 * Space Complexity: O(1) as we only use a few extra variables.
 */
export function removeNthFromEndTwoPass(head: ListNode | null, n: number): ListNode | null {
    const dummy = new ListNode(0);
    dummy.next = head;

    let length = 0;
    let current: ListNode | null = head;
    // First pass: Calculate the length of the list
    while (current !== null) {
        length++;
        current = current.next;
    }

    // Calculate the position of the node to remove from the beginning (0-indexed)
    // We want to stop at the node *before* the one to be removed.
    const nodeToRemoveFromStart = length - n;

    current = dummy; // Reset current to dummy to start second pass
    // Second pass: Traverse to the node *before* the one to remove
    for (let i = 0; i < nodeToRemoveFromStart; i++) {
        current = current!.next;
    }

    // current.next is the node to be removed
    // Bypass it by setting current.next to current.next.next
    if (current && current.next) {
        current.next = current.next.next;
    }

    return dummy.next; // Return the head of the modified list
}